File: backend/app/api/v1/knowledge.py
"""Knowledge Base 的文本摄取、发布与检索 API。"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, HTTPException

from app.api.deps import DB, CurrentUser
from app.models.knowledge import (
    KnowledgeBase,
    KnowledgeBaseMember,
    KnowledgeDocument,
    KnowledgeSource,
)
from app.schemas.knowledge import (
    KnowledgeBaseCreate,
    KnowledgeBaseResponse,
    KnowledgeCitation,
    KnowledgeDocumentResponse,
    KnowledgeMemberGrant,
    KnowledgeMemberResponse,
    KnowledgeSearchRequest,
    KnowledgeSourceWithDocumentsResponse,
    KnowledgeTextSourceCreate,
    KnowledgeTextVersionCreate,
)
from app.services.knowledge_service import (
    KnowledgeNotFoundError,
    KnowledgePermissionError,
    create_knowledge_base,
    create_text_source,
    create_text_version,
    grant_member,
    list_knowledge_bases,
    list_knowledge_sources,
    publish_document,
    search_knowledge,
)

router = APIRouter(prefix="/knowledge-bases", tags=["knowledge-bases"])


@router.post("", response_model=KnowledgeBaseResponse, status_code=201)
async def create_base(
    request: KnowledgeBaseCreate, current_user: CurrentUser, db: DB
) -> KnowledgeBase:
    """创建一个当前用户拥有的知识库。"""

    knowledge_base = await create_knowledge_base(
        owner_id=current_user.id, name=request.name, space_id=request.space_id, db=db
    )
    await db.commit()
    await db.refresh(knowledge_base)
    return knowledge_base


@router.get("", response_model=list[KnowledgeBaseResponse])
async def get_bases(current_user: CurrentUser, db: DB) -> list[KnowledgeBase]:
    """仅列出当前用户有空间 ACL 的知识库。"""

    return await list_knowledge_bases(user_id=current_user.id, db=db)


@router.get(
    "/{knowledge_base_id}/sources", response_model=list[KnowledgeSourceWithDocumentsResponse]
)
async def get_sources(
    knowledge_base_id: uuid.UUID, current_user: CurrentUser, db: DB
) -> list[KnowledgeSource]:
    """列出可访问知识库的来源及其版本。"""

    try:
        return await list_knowledge_sources(
            knowledge_base_id=knowledge_base_id, user_id=current_user.id, db=db
        )
    except KnowledgeNotFoundError as error:
        raise HTTPException(status_code=404, detail="KNOWLEDGE_BASE_NOT_FOUND") from error


@router.put("/{knowledge_base_id}/members", response_model=KnowledgeMemberResponse)
async def put_member(
    knowledge_base_id: uuid.UUID, request: KnowledgeMemberGrant, current_user: CurrentUser, db: DB
) -> KnowledgeBaseMember:
    """由所有者为一个用户设置知识库成员角色。"""

    try:
        member = await grant_member(
            knowledge_base_id=knowledge_base_id,
            owner_id=current_user.id,
            user_id=request.user_id,
            role=request.role,
            db=db,
        )
    except KnowledgeNotFoundError as error:
        raise HTTPException(status_code=404, detail="KNOWLEDGE_BASE_NOT_FOUND") from error
    except KnowledgePermissionError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error
    await db.commit()
    await db.refresh(member)
    return member


@router.post(
    "/{knowledge_base_id}/sources/text", response_model=KnowledgeDocumentResponse, status_code=201
)
async def create_source(
    knowledge_base_id: uuid.UUID,
    request: KnowledgeTextSourceCreate,
    current_user: CurrentUser,
    db: DB,
) -> KnowledgeDocument:
    """创建文本来源并构建未公开的初始版本。"""

    try:
        _, document = await create_text_source(
            knowledge_base_id=knowledge_base_id, user_id=current_user.id, request=request, db=db
        )
    except KnowledgeNotFoundError as error:
        raise HTTPException(status_code=404, detail="KNOWLEDGE_BASE_NOT_FOUND") from error
    except KnowledgePermissionError as error:
        raise HTTPException(status_code=403, detail="KNOWLEDGE_BASE_WRITE_FORBIDDEN") from error
    except ValueError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error
    await db.commit()
    await db.refresh(document)
    return document


@router.post(
    "/{knowledge_base_id}/sources/{source_id}/documents/text",
    response_model=KnowledgeDocumentResponse,
    status_code=201,
)
async def create_source_version(
    knowledge_base_id: uuid.UUID,
    source_id: uuid.UUID,
    request: KnowledgeTextVersionCreate,
    current_user: CurrentUser,
    db: DB,
) -> KnowledgeDocument:
    """为现有来源构建待发布文本版本。"""

    try:
        document = await create_text_version(
            knowledge_base_id=knowledge_base_id,
            source_id=source_id,
            user_id=current_user.id,
            content=request.content,
            db=db,
        )
    except KnowledgeNotFoundError as error:
        raise HTTPException(status_code=404, detail="KNOWLEDGE_SOURCE_NOT_FOUND") from error
    except KnowledgePermissionError as error:
        raise HTTPException(status_code=403, detail="KNOWLEDGE_BASE_WRITE_FORBIDDEN") from error
    except ValueError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error
    await db.commit()
    await db.refresh(document)
    return document


@router.post(
    "/{knowledge_base_id}/sources/{source_id}/documents/{document_id}/publish",
    response_model=KnowledgeDocumentResponse,
)
async def publish_source_version(
    knowledge_base_id: uuid.UUID,
    source_id: uuid.UUID,
    document_id: uuid.UUID,
    current_user: CurrentUser,
    db: DB,
) -> KnowledgeDocument:
    """原子地将已构建版本切换为可检索版本。"""

    try:
        document = await publish_document(
            knowledge_base_id=knowledge_base_id,
            source_id=source_id,
            document_id=document_id,
            user_id=current_user.id,
            db=db,
        )
    except KnowledgeNotFoundError as error:
        raise HTTPException(status_code=404, detail="KNOWLEDGE_DOCUMENT_NOT_FOUND") from error
    except KnowledgePermissionError as error:
        raise HTTPException(status_code=403, detail="KNOWLEDGE_BASE_WRITE_FORBIDDEN") from error
    await db.commit()
    await db.refresh(document)
    return document


@router.post("/{knowledge_base_id}/search", response_model=list[KnowledgeCitation])
async def search_base(
    knowledge_base_id: uuid.UUID, request: KnowledgeSearchRequest, current_user: CurrentUser, db: DB
) -> list[KnowledgeCitation]:
    """检索当前用户可访问的已发布来源片段与引用位置。"""

    return await search_knowledge(
        knowledge_base_id=knowledge_base_id,
        user_id=current_user.id,
        query=request.query,
        limit=request.limit,
        db=db,
    )

File: packages/core/src/api/knowledge.ts
import type {
  CreateKnowledgeTextSourceInput,
  KnowledgeBase,
  KnowledgeCitation,
  KnowledgeDocument,
  KnowledgeSource,
} from '@yuanai/types'

import { apiClient } from './client.js'

/** 创建知识库所需的基础元数据。 */
export interface CreateKnowledgeBaseInput {
  name: string
  spaceId?: string
}

/** 对知识库执行测试检索的输入。 */
export interface KnowledgeSearchInput {
  query: string
  limit?: number
}

/** 列出当前用户可访问的知识库。 */
export async function listKnowledgeBases(): Promise<KnowledgeBase[]> {
  return (await apiClient.get<KnowledgeBase[]>('/knowledge-bases')).data
}

/** 创建一个当前用户拥有的知识库。 */
export async function createKnowledgeBase(input: CreateKnowledgeBaseInput): Promise<KnowledgeBase> {
  return (await apiClient.post<KnowledgeBase>('/knowledge-bases', input)).data
}

/** 为知识库创建带有首个暂存版本的纯文本来源。 */
export async function createKnowledgeTextSource(
  knowledgeBaseId: string,
  input: CreateKnowledgeTextSourceInput
): Promise<KnowledgeDocument> {
  return (
    await apiClient.post<KnowledgeDocument>(
      `/knowledge-bases/${knowledgeBaseId}/sources/text`,
      input
    )
  ).data
}

/** 列出一个知识库现有来源及其版本。 */
export async function listKnowledgeSources(knowledgeBaseId: string): Promise<KnowledgeSource[]> {
  return (await apiClient.get<KnowledgeSource[]>(`/knowledge-bases/${knowledgeBaseId}/sources`))
    .data
}

/** 原子发布一个已构建的知识库来源版本。 */
export async function publishKnowledgeDocument(
  knowledgeBaseId: string,
  sourceId: string,
  documentId: string
): Promise<KnowledgeDocument> {
  return (
    await apiClient.post<KnowledgeDocument>(
      `/knowledge-bases/${knowledgeBaseId}/sources/${sourceId}/documents/${documentId}/publish`
    )
  ).data
}

/** 检索当前用户可访问的已发布知识库片段。 */
export async function searchKnowledgeBase(
  knowledgeBaseId: string,
  input: KnowledgeSearchInput
): Promise<KnowledgeCitation[]> {
  return (
    await apiClient.post<KnowledgeCitation[]>(`/knowledge-bases/${knowledgeBaseId}/search`, input)
  ).data
}
